### PR TITLE
Workspaces: Keep already-set owner annotations

### PR DESCRIPTION
**Pull request by davidfestal**
## Summary

This PR allows keeping an already-set `owner` annotation on workspace creation, when it has been set by a `system:masters` user.

## Related issue(s)

This is related to required cleanup of obsolete parts in the `workspaces` virtual workspace to prepare for the Home workspace implementation.

### CODE REVIEW

**Review by stevekuznetsov — CHANGES_REQUESTED** [review]
I don't think we should be allowing people to provide this annotation - it's a system internal and can change at any time, getting clients to depend on it is bad. Why not impersonate or create a virtual endpoint for this?

**Comment by davidfestal** [comment]
> I don't think we should be allowing people to provide this annotation - it's a system internal and can change at any time, getting clients to depend on it is bad. Why not impersonate or create a virtual endpoint for this?

It's not allowing *people*, but allowing users in the `system:masters` group, especially allowing the `workspaces` virtual workspace to set this annotation.
@sttts proposed this as a way to avoid impersonation from the `workspaces` virtual workspace side.

**Comment by stevekuznetsov** [comment]
IMHO this will come back to bite us. This is not an API. At the minimum if we must do this, please parse the existing annotation and make sure it is valid JSON and deserializes into user info as expected.

**Comment by davidfestal** [comment]
>At the minimum if we must do this, please parse the existing annotation and make sure it is valid JSON and deserializes into user info as expected.
Sure, I'll do that @stevekuznetsov

**Comment by davidfestal** [comment]
> > At the minimum if we must do this, please parse the existing annotation and make sure it is valid JSON and deserializes into user info as expected.
> > Sure, I'll do that @stevekuznetsov

Done in commit https://github.com/kcp-dev/kcp/pull/1384/commits/3c46b23c3ac7f8f4dc28659db563f13744d2263e

**Comment by sttts** [comment]
It's the workspace VW. There cannot be anything more internal qualified to set internal workspace annotations than this.

File: pkg/admission/clusterworkspacetypeexists/admission.go [hunk 1]
```diff
@@ -121,7 +123,15 @@ func (o *clusterWorkspaceTypeExists) Admit(ctx context.Context, a admission.Attr
...
 		}
-		cw.Annotations[tenancyv1alpha1.ClusterWorkspaceOwnerAnnotationKey] = string(rawInfo)
+
+		if sets.NewString(currentUser.GetGroups()...).Has(user.SystemPrivilegedGroup) {
+			// Don't override the already existing owner annotation if set by a system:masters user
+			if _, exists := cw.Annotations[tenancyv1alpha1.ClusterWorkspaceOwnerAnnotationKey]; !exists {
+				cw.Annotations[tenancyv1alpha1.ClusterWorkspaceOwnerAnnotationKey] = string(rawInfo)
+			}
+		} else {
+			cw.Annotations[tenancyv1alpha1.ClusterWorkspaceOwnerAnnotationKey] = string(rawInfo)
```
[the remark this answers is not in the document]
Comment (s-urbaniak) [reply pkg/admission/clusterworkspacetypeexists/admission.go:138]: https://github.com/kcp-dev/kcp/blob/main/pkg/admission/reservedmetadata/admission.go

**Comment by sttts** [comment]
1. please move this into its own admission plugin. Having it there is wrong, and super hard to understand what's going on.
2. the owner MUST be verified in the corresponding Validate method. If it is not, please add it.
3. if validate checks the correctness, then @s-urbaniak can exclude this annotation from those that are protected.

🔴 These are blockers for moving forward.

**Comment by davidfestal** [comment]
@sttts Did you already fixed this in another PR ? Should this one be closed ?

**Comment by sttts** [comment]
> @sttts Did you already fixed this in another PR ? Should this one be closed ?

yes. Will resume work on monday.

**Pull request closed without merging**